### PR TITLE
[WIP] MedlineImporterTest

### PR DESCRIPTION
**Pull request by bruehldev**
New PR started because the old one was on a wrong branch.

MedlineImporTest + Testfile added

Couldn't cover two exceptions in importEntries.

### CODE REVIEW

**Comment by oscargus** [comment]
BibtexEntry is called BibEntry now. Not sure how the CircleCI tests passes, but that is the reason that the Travis CI tests won't.

**Comment by bruehldev** [comment]
I'm afraid i cant use the BibEntry. Somehow i can't use the libary ("BibEntry cannot be resolved to a type"), but i fetched my upstream.

**Comment by oscargus** [comment]
Do you have a package net.sf.jabref.model.entry.BibEntry? model/entry/BibEntry.java
I'm more surprised that you still have BibtexEntry.java.

**Comment by bruehldev** [comment]
I updated the branch with BibEntry. We will solve the package problem with our  tutor. Thanks Oscar

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import org.junit.Test;
+import org.junit.Assert;
+
+/**
+ * @author Daniel Mair/Bruehl
+ *
+ */
+public class MedlineImporterTest {
+
+    MedlineImporter medlineImp;
```
Comment (simonharrer) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: Do not use abbreviations. Either use `medlineImporter` or `importer` as the context is clear by being in the `MedlineImporterTest` class.

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                Assert.assertFalse(medlineImp.isRecognizedFormat(is));
+            }
+        }
+    }
+
+    @Test
+    public void testImportEntries() throws IOException {
+        try (InputStream is = MedlineImporter.class.getResourceAsStream("MedlineImporterTest.xml")) {
+            List<BibEntry> entries = medlineImp.importEntries(is, new OutputPrinterToNull());
+            Assert.assertEquals(1, entries.size());
```
Comment (simonharrer) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: This is not a good assertion, you need to verify as well all the fields of the BibEntry.

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+ */
+public class MedlineImporterTest {
+
+    MedlineImporter medlineImp;
+
+    @Before
+    public void setUp() throws Exception {
+        if (Globals.prefs == null) {
+            Globals.prefs = JabRefPreferences.getInstance();
+        }
```
Comment (simonharrer) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: You do not need the if-condition here, just set the `Globals.prefs`

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        try (InputStream is = MedlineImporter.class.getResourceAsStream("CopacImporterTest1.txt")) {
+            // Trigger a catchblock
+            List<BibEntry> entries = medlineImp.importEntries(is, new OutputPrinterToNull());
+            //Assert.fail("Error during XML parsing");
+        }
+    }
+
+    @Test
+    public void testGetFormatName() {assertEquals("testGetFormatName failed", "Medline", medlineImp.getFormatName());
+    }
```
Comment (simonharrer) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: Mind the formatting, please.

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Test
+    public void testGetCLIId() {
+        assertEquals("testGetFormatName failed", "medline", medlineImp.getCLIId());
```
Comment (simonharrer) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: Beware of copy-and-paste errors, as the message here is completely wrong. A message of an assert should add additional information, which in this case, it does not, it adds misleading information.

**Comment by mairdl** [comment]
For now I took care of the smaller changes, more changes coming soon

**Comment by boceckts** [comment]
Added more test files and added testing against bibtex files.

**Comment by koppor** [comment]
@boceckts Is this ready for review? Then please add this label?

**Comment by boceckts** [comment]
Testcoverage is at only ~80% so far for MedlineImporter and at ~90% for MedlineHandler

**Comment by stefan-kolb** [comment]
What is the status here? DL was beginning of match.

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    @Test
+    public void testImportEntryWithArticleID() throws IOException {
+        checkImportEntriesAgainstBibFileEntries("MedlineImporterArticleIDTest.xml",
+                "MedlineImporterArticleIDTestBib.bib");
+        try (InputStream shouldBeIs = MedlineImporter.class.getResourceAsStream("MedlineImporterArticleIDTest.xml");
+                InputStream bibIn = MedlineImporter.class.getResourceAsStream("MedlineImporterArticleIDTestBib.bib")) {
+            List<BibEntry> entries = medlineImporter.importEntries(shouldBeIs, new OutputPrinterToNull());
+            Assert.assertEquals(1, entries.size());
```
Comment (tschechlovdev) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: this assertion is not necessary, it is done in BibtexEntryAssert(InputStream, List<BibEntry>)

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            BibEntry entry = entries.get(0);
+            BibtexEntryAssert.assertEquals(bibIn, entry);
```
Comment (tschechlovdev) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: use BibtexEntryAssert.assertEquals(InputStream, List<BibEntry>)

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                "MedlineImporterMinimalEntryTestBib.bib");
+    }
+
+    private void checkImportEntriesAgainstBibFileEntries(String medlineFile, String bibTexFile) throws IOException {
+        try (InputStream is = MedlineImporter.class.getResourceAsStream(medlineFile);
+                InputStream nis = MedlineImporter.class.getResourceAsStream(bibTexFile)) {
+            List<BibEntry> entries = medlineImporter.importEntries(is, new OutputPrinterToNull());
+            Assert.assertNotNull(entries);
+            Assert.assertEquals(1, entries.size());
+            BibtexEntryAssert.assertEquals(nis, entries.get(0));
```
Comment (tschechlovdev) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: See above. 
Assert.assertNotNull is not necessary as well if you use `assertEquals(InputStram, List<BibEntry>)`

**Comment by koppor** [comment]
Please rebase everything on master and create on commit. Description is here: https://github.com/JabRef/jabref/wiki/Tools#rebase-everything-as-one-commit-on-master Please use `git gui` for your commit. Please do not delete other test files.

There should no commits handling the InspecImporter in this pull request. Inspec has been handled in #356

**Comment by mairdl** [comment]
I am running into a problem in the "GrammerBasedSearchRule.java" it is missing some required imports. The missing one are "net.sf.jabref.search.SearchBaseVisitor", "net.sf.jabref.search.SearchLexer" and "net.sf.jabref.search.SearchBaseParser"

**Comment by tobiasdiez** [comment]
@mairdl You have to run `gradlew generateSource`.

**Comment by mairdl** [comment]
Now it seems to work, thank you.

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    public void testGetCLIId() {
+        assertEquals("medline", medlineImporter.getCLIId());
+    }
+
+    @Test
+    public void testFetchMedline() {
+        List<BibEntry> validMedLine = medlineImporter.fetchMedline("1", new OutputPrinterToNull());
+        List<BibEntry> emptyMedLine = medlineImporter.fetchMedline("zzzzzz", new OutputPrinterToNull());
+        assertNotNull(validMedLine);
+        assertEquals(0, emptyMedLine.size());
```
Comment (tschechlovdev) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: Use assertEquals(Collections.EMPTY_LIST, emptyMedline) instead

File: src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Test
+    public void testIsRecognizedFormatEmptyFile() throws IOException {
+        Assert.assertFalse(medlineImporter.isRecognizedFormat(emptyFileStream));
+    }
+
+    @Test
+    public void testImportMalformedEntry() throws IOException {
+        try (InputStream is = MedlineImporter.class.getResourceAsStream("MedlineImporterMalformedEntryTest.xml")) {
+            medlineImporter.importEntries(is, new OutputPrinterToNull());
```
Comment (tschechlovdev) [inline src/test/java/net/sf/jabref/importer/fileformat/MedlineImporterTest.java]: why is here no assertion?

File: src/test/resources/net/sf/jabref/importer/fileformat/MedlineImporterArticleIDTestBib.bib [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
...ial to optimize survival. We evaluated the utility of a multiplex PCR test in differentiating possible etiologies of meningitis. Cerebrospinal fluid (CSF) from 69 HIV-infected Ugandan adults with meningitis was collected at diagnosis (n=51) and among persons with cryptococcal meningitis during therapeutic lumbar punctures (n=68). Cryopreserved CSF specimens were analyzed with BioFire FilmArray® Meningitis/Encephalitis panel, which targets 17 pathogens. The panel detected Cryptococcus in the CSF of patients diagnosed with a first episode of cryptococcal meningitis by fungal culture with 100\% sensitivity and specificity and differentiated between fungal relapse and paradoxical immune reconstitution inflammatory syndrome in recurrent episodes. A negative FilmArray result was predictive of CSF sterility on follow-up lumbar punctures for cryptococcal meningitis. EBV was frequently detected in this immunosuppressed population (n=45). Other pathogens detected included: cytomegalovirus (n=2), varicella zoster virus (n=2), human herpes virus 6 (n=1), and Streptococcus pneumoniae (n=1). The FilmArray Meningitis/Encephalitis panel offers a promising platform for rapid meningitis diagnosis.},
```
Comment (boceckts) [inline src/test/resources/net/sf/jabref/importer/fileformat/MedlineImporterArticleIDTestBib.bib]: Please use a dummy abstract instead of an actual one

File: src/test/resources/net/sf/jabref/importer/fileformat/MedlineImporterTestArticleID.bib [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
...ial to optimize survival. We evaluated the utility of a multiplex PCR test in differentiating possible etiologies of meningitis. Cerebrospinal fluid (CSF) from 69 HIV-infected Ugandan adults with meningitis was collected at diagnosis (n=51) and among persons with cryptococcal meningitis during therapeutic lumbar punctures (n=68). Cryopreserved CSF specimens were analyzed with BioFire FilmArray® Meningitis/Encephalitis panel, which targets 17 pathogens. The panel detected Cryptococcus in the CSF of patients diagnosed with a first episode of cryptococcal meningitis by fungal culture with 100\% sensitivity and specificity and differentiated between fungal relapse and paradoxical immune reconstitution inflammatory syndrome in recurrent episodes. A negative FilmArray result was predictive of CSF sterility on follow-up lumbar punctures for cryptococcal meningitis. EBV was frequently detected in this immunosuppressed population (n=45). Other pathogens detected included: cytomegalovirus (n=2), varicella zoster virus (n=2), human herpes virus 6 (n=1), and Streptococcus pneumoniae (n=1). The FilmArray Meningitis/Encephalitis panel offers a promising platform for rapid meningitis diagnosis.},
```
Comment (tschechlovdev) [inline src/test/resources/net/sf/jabref/importer/fileformat/MedlineImporterTestArticleID.bib]: use a dummy abstract

**Comment by mairdl** [comment]
The test are now parameterized as suggested by @tschechlovdev. Also I removed the copyrighted abstract.

**Comment by simonharrer** [comment]
Please fix codacy issues.

File: src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java [hunk 13]
```diff
@@ -646,7 +646,14 @@ public static String expandAuthorInitials(String name) {
...
                     if (j == 1) {
-                        sb.append(Author.addDotIfAbbreviation(names[j]));
+                        String[] splittedNames = names[j].split(" ");
+                        if (splittedNames.length > 0) {
+                            sb.append(Author.addDotIfAbbreviation(splittedNames[0]));
+                        }
+                        for (int k = 1; k < splittedNames.length; k++) {
+                            sb.append(' ');
+                            sb.append(Author.addDotIfAbbreviation(splittedNames[k]));
+                        }
```
Comment (simonharrer) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: If you split, you should join afterwards using `String.join(" ", elems);`

Comment (mairdl) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: I am not sure if that's needed here, because the whole method `expandAuthorInitials()` doesn't use it all. This might be the case because the string that gets splitted is used to build a new string via `StringBuilder`.

**Comment by mairdl** [comment]
Some tests outside the `MedlinImporterTest` and `MedlineImporterTestFiles` are failing. 
The class is the `ContainsAndRegexBasedSearchRuleDescriberTest`  class and the `GrammarBasedSearchRuleDescriberTest` class. The expected value is a `String` in english but the actual `String` is written in german.

File: src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java (hunk 13 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: Can you please shortly outline why this change is necessary?

Comment (mairdl) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: A test case revealed this issue. The importent lines were these:
 `<Author>
                    <CollectiveName>ASTRO-CM Team</CollectiveName>
                </Author>` 
Bewteen ASTRO-CM and Team it would add an additional whitespace. With this change the whitespace won't be added.

**Comment by mairdl** [comment]
Codacy issues should be fixed now.

File: src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java (hunk 13 above)
Comment (koppor) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: @mairdl The code you provided is the same as

```java
List<String> innerNameList = Arrays.asList(splittedNames);
String innerNames  innerNameList.stream().map(n -> Author.addDotIfAbbreviation(n)).collect(Collectors.joining(" "));
sb.append(innerNames);
```

(untested, please double check if it really works :innocent:)

This is much shorter than your code, probably more readable (if one is used to Java8) and less error prone.

See also http://stackoverflow.com/a/22577565/873282

Comment (mairdl) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: The first verson passes all test. Can't say that for you 2nd example.

I agree  that it's more readable, shall i change it? The code i used was inspired by the `else `part of this method.

Comment (simonharrer) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: For simplicity, I would probably use the StringJoiner class without streams.

Comment (koppor) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: Ah, I see. According to [the JAvaDoc of StringJoiner](https://docs.oracle.com/javase/8/docs/api/java/util/StringJoiner.html), one can also transform the inputs (achieved by `map` in my proposal). Since the iteration on the split result (an array) is easy via an foreach, this should easily work without some if checking for non-emptyness (`splittedNames.length > 0`) etc.

**Comment by mairdl** [comment]
I'd like to move the issue #1267 to a seperate PR, to have a cleaner cut

**Comment by matthiasgeiger** [comment]
So you have created a test for importing two or more entries which fails?

**Comment by mairdl** [comment]
No, but @koppor wants tests even if it not fails

**Comment by matthiasgeiger** [comment]
Yes. And this tests should be included here, I think. 

If the tests fail, the failing tests in conjunction with the bugfix can be in another PR.

**Comment by koppor** [comment]
The failed tests can be included here with `@Ignore("Will be fixed in a future PR")` so that there is no huzzle with git, can't they?

**Comment by matthiasgeiger** [comment]
Yes, of course and this is an even better solution. :wink:

File: src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java (hunk 13 above)
Comment (mairdl) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: The problem here is that atleast to my experience, that a `StringJoiner` and `StringBuilder` or another `StringJoiner` has to be used. This is due to the reason that it first has to put the last name and first name together with a `", "` after that it should add that `"last name, first name"` with a `" and "` to the `String` that gets returned. I don't know if that increases simplicity. I suggest using one of koppor's propasals.

PS: Sorry for the stall, I had an operation this week and was resting.

**Comment by mairdl** [comment]
After some testing i came to the conclusion that this bug/issue is not a problem in the current development version. At the moment I am struggeling to coming up with the right test case. At the moment I am thinking of using a `file.bib` wit some pubmed entries and then importing it with the RISimporter and checking if it works properly. But that seems a bit unfitting for the MedlineImporter.

**Comment by simonharrer** [comment]
What is the current status of this PR?

**Comment by mairdl** [comment]
waiting for some feedback regardnig the issue #1267 and my comment and also on comments from me regarding the use of a `StringJoiner`

File: src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java (hunk 13 above)
Comment (simonharrer) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: The StringBuilder has nothing to do with the StringJoiner in that case. Just use this solution:

```java
String innerNames = Arrays.stream(names[j].split(" "))
    .map(Author::addDotIfAbbreviation)
    .collect(Collectors.joining(" "));
sb.append(innerNames);
```

Comment (mairdl) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: This leads to some errors in the `StringUtilTest.java` because the suffix gets ignored.

**Comment by mairdl** [comment]
@simonharrer your suggested version leads to some problems

File: src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java (hunk 13 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: We decided to completely remove the "expand author initials" functionality in #1373.

**Comment by tobiasdiez** [comment]
Since this PR blocks #1207, I would prefer that we merge this as quickly as possible. So just ignore the issue #1267 for the moment.

**Comment by simonharrer** [comment]
@mairdl hm, then revert it to the state without my suggestion, and rebase. We try to merge this in quickly as it blocks #1207

**Comment by mairdl** [comment]
ok i will revert it asap

**Comment by mairdl** [comment]
I reverted the changes of StringUtil and added a ignored label to the tests that would fail.

**Comment by simonharrer** [comment]
Thanks. Please open a new PR then for the fix of the failing test. :)
